File: packages/moreau-cuda/examples/cpp/large_qp_batched.cpp
#include "moreau/solver/solver.hpp"
#include <iostream>
#include <vector>
#include <random>
#include <algorithm>
#include <limits>

using namespace moreau;

int main(int argc, char** argv) {
    // Usage: large_qp_batched [batch_size] [n] [m] [density]

    // Default problem size
    int n = 5000;
    int m = 2500;
    int batchSize = 128;
    double density = 0.05;  // 5% density by default
    bool skip_solve = false;
    bool use_schur = false;

    // Parse arguments
    int pos_arg = 0;
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--skip-solve") {
            skip_solve = true;
        } else if (arg == "--schur") {
            use_schur = true;
        } else {
            // Positional arguments: batch_size, n, m, density
            if (pos_arg == 0) batchSize = std::atoi(argv[i]);
            else if (pos_arg == 1) n = std::atoi(argv[i]);
            else if (pos_arg == 2) m = std::atoi(argv[i]);
            else if (pos_arg == 3) density = std::atof(argv[i]);
            pos_arg++;
        }
    }

    std::cout << "Creating large batched QP problem:\n";
    std::cout << "  Variables: " << n << "\n";
    std::cout << "  Constraints: " << m << "\n";
    std::cout << "  Batch size: " << batchSize << "\n";
    std::cout << "  Target density: " << (density * 100) << "%\n";

    // Random number generator
    std::mt19937 rng(42);
    std::uniform_real_distribution<double> dist(0.1, 2.0);

    // P matrix: diagonal with random positive values (n x n)
    // Same structure for all batches
    std::vector<int64_t> P_rowOffsets(n + 1);
    std::vector<int64_t> P_colIndices(n);

    for (int i = 0; i < n; i++) {
        P_rowOffsets[i] = i;
        P_colIndices[i] = i;
    }
    P_rowOffsets[n] = n;
    int nnzP = n;

    // Generate different P values for each batch
    std::vector<double> P_values(nnzP * batchSize);
    for (int batch = 0; batch < batchSize; batch++) {
        for (int i = 0; i < n; i++) {
            P_values[batch * nnzP + i] = dist(rng);
        }
    }

    // q vector: different random linear cost for each batch
    std::vector<double> q_values(n * batchSize);
    std::uniform_real_distribution<double> q_dist(-1.0, 1.0);
    for (int batch = 0; batch < batchSize; batch++) {
        for (int i = 0; i < n; i++) {
            q_values[batch * n + i] = q_dist(rng);
        }
    }

    // A matrix: sparse random (m x n), with specified density
    // Same structure for all batches
    std::vector<int64_t> A_rowOffsets;
    std::vector<int64_t> A_colIndices;

    A_rowOffsets.push_back(0);
    int64_t nnz = 0;

    // Compute entries per row from density
    int entries_per_row = std::max(1, std::min(n, static_cast<int>(density * n + 0.5)));

    for (int i = 0; i < m; i++) {
        // Sample random columns without replacement
        std::vector<int> all_cols(n);
        std::iota(all_cols.begin(), all_cols.end(), 0);
        std::shuffle(all_cols.begin(), all_cols.end(), rng);

        // Take first entries_per_row columns
        std::vector<int> cols(all_cols.begin(), all_cols.begin() + entries_per_row);
        std::sort(cols.begin(), cols.end());

        for (int col : cols) {
            A_colIndices.push_back(col);
            nnz++;
        }

        A_rowOffsets.push_back(nnz);
    }

    int nnzA = nnz;
    std::cout << "A matrix sparsity: " << nnzA << " nonzeros ("
              << (100.0 * nnzA / (n * m)) << "% dense)\n";

    // Generate different A values for each batch
    std::vector<double> A_values(nnzA * batchSize);
    for (int batch = 0; batch < batchSize; batch++) {
        for (int i = 0; i < nnzA; i++) {
            A_values[batch * nnzA + i] = dist(rng);
        }
    }

    // b vector (RHS) - different for each batch
    std::vector<double> b_values(m * batchSize);
    std::uniform_real_distribution<double> b_dist(-2.0, 2.0);
    for (int batch = 0; batch < batchSize; batch++) {
        for (int i = 0; i < m; i++) {
            b_values[batch * m + i] = b_dist(rng);
        }
    }

    // Cone configuration: all nonnegative (simple box constraints)
    Cones cones;
    cones.numZeroCones = 0;
    cones.numNonnegCones = m;  // All constraints are x >= 0
    cones.numExpCones = 0;
    cones.numSocCones = 0;
    cones.numPowerCones = 0;

    // Settings
    Settings settings;
    settings.ipm.equilibrationSettings.enable = true;
    settings.maxIter = 50;
    settings.verbose = false;  // Enable to see detailed timing breakdown

    std::cout << "Creating solver...\n";

    // Measure constructor time
    cudaEvent_t ctor_start, ctor_stop;
    cudaEventCreate(&ctor_start);
    cudaEventCreate(&ctor_stop);

    cudaEventRecord(ctor_start);
    CompiledSolver* solver_ptr;
    {
        // Create solver
        solver_ptr = new CompiledSolver(
            n, m, batchSize,
            P_rowOffsets.data(), P_colIndices.data(), nnzP,
            A_rowOffsets.data(), A_colIndices.data(), nnzA,
            cones,
            settings
        );
    }
    CompiledSolver& solver = *solver_ptr;
    cudaEventRecord(ctor_stop);
    cudaEventSynchronize(ctor_stop);

    float ctor_ms = 0;
    cudaEventElapsedTime(&ctor_ms, ctor_start, ctor_stop);
    std::cout << "Constructor time: " << ctor_ms << " ms\n";

    cudaEventDestroy(ctor_start);
    cudaEventDestroy(ctor_stop);

    // Allocate device memory for problem data
    double *d_P_values, *d_A_values, *d_q, *d_b;
    cudaMalloc(&d_P_values, sizeof(double) * nnzP * batchSize);
    cudaMalloc(&d_A_values, sizeof(double) * nnzA * batchSize);
    cudaMalloc(&d_q, sizeof(double) * n * batchSize);
    cudaMalloc(&d_b, sizeof(double) * m * batchSize);

    // Copy to device
    cudaMemcpy(d_P_values, P_values.data(), sizeof(double) * nnzP * batchSize, cudaMemcpyHostToDevice);
    cudaMemcpy(d_A_values, A_values.data(), sizeof(double) * nnzA * batchSize, cudaMemcpyHostToDevice);
    cudaMemcpy(d_q, q_values.data(), sizeof(double) * n * batchSize, cudaMemcpyHostToDevice);
    cudaMemcpy(d_b, b_values.data(), sizeof(double) * m * batchSize, cudaMemcpyHostToDevice);

    std::cout << "\nSolving " << batchSize << " problems...\n";

    // Benchmark the solve
    cudaEvent_t start, stop;
    cudaEventCreate(&start);
    cudaEventCreate(&stop);

    // Measure setup separately
    cudaEvent_t setup_start, setup_stop;
    cudaEventCreate(&setup_start);
    cudaEventCreate(&setup_stop);

    cudaEventRecord(setup_start);
    {
        solver.setup(d_P_values, d_A_values);
    }
    cudaEventRecord(setup_stop);
    cudaEventSynchronize(setup_stop);

    float setup_ms = 0;
    cudaEventElapsedTime(&setup_ms, setup_start, setup_stop);

    // Measure solve separately
    cudaEventRecord(start);
    if (!skip_solve) {
        solver.solve(d_q, d_b);
    } else {
        std::cout << "SKIPPED solve() call\n";
    }
    cudaEventRecord(stop);

    cudaEventSynchronize(stop);
    float solve_ms = 0;
    cudaEventElapsedTime(&solve_ms, start, stop);

    float total_ms = setup_ms + solve_ms;

    std::cout << "\n=== BENCHMARK ===" << std::endl;
    std::cout << "Setup time: " << setup_ms << " ms" << std::endl;
    std::cout << "Solve time: " << solve_ms << " ms" << std::endl;
    std::cout << "Total time: " << total_ms << " ms" << std::endl;
    std::cout << "Time per problem (solve only): " << (solve_ms / batchSize) << " ms" << std::endl;
    std::cout << "Time per problem (total): " << (total_ms / batchSize) << " ms" << std::endl;
    std::cout << "Throughput (solve only): " << (1000.0 * batchSize / solve_ms) << " problems/second" << std::endl;

    cudaEventDestroy(setup_start);
    cudaEventDestroy(setup_stop);

    cudaEventDestroy(start);
    cudaEventDestroy(stop);

    // Print solution statistics
    std::vector<double> x_host(n * batchSize);
    cudaMemcpy(x_host.data(), solver.solution.x.data(),
               sizeof(double) * n * batchSize, cudaMemcpyDeviceToHost);
    (void)use_schur;  // unused for now

    std::cout << "\n=== SOLUTION (first problem, first 10 values) ===\n";
    std::cout << "x = [";
    for (int i = 0; i < std::min(10, n); i++) {
        std::cout << x_host[i];
        if (i < std::min(10, n)-1) std::cout << ", ";
    }
    std::cout << ", ...]\n";

    // Print final info for all batches
    std::vector<double> cost_primal(batchSize);
    solver.info.cost_primal.gpuToCpu(cost_primal.data());

    std::vector<int32_t> status_vals(batchSize);
    cudaMemcpy(status_vals.data(), solver.info.status_device,
               sizeof(int32_t) * batchSize, cudaMemcpyDeviceToHost);

    std::cout << "\n=== BATCH RESULTS ===\n";
    std::cout << "Total iterations: " << solver.info.iterations << "\n";

    int num_solved = 0;
    int num_almost_solved = 0;
    int num_failed = 0;

    for (int batch = 0; batch < batchSize; batch++) {
        if (status_vals[batch] == 1) num_solved++;
        else if (status_vals[batch] == 6) num_almost_solved++;
        else num_failed++;
    }

    std::cout << "Solved: " << num_solved << " / " << batchSize << "\n";
    std::cout << "Almost solved: " << num_almost_solved << " / " << batchSize << "\n";
    std::cout << "Failed: " << num_failed << " / " << batchSize << "\n";

    if (batchSize <= 10) {
        std::cout << "\n=== PER-BATCH DETAILS ===\n";
        for (int batch = 0; batch < batchSize; batch++) {
            std::cout << "Batch " << batch << ": objective=" << cost_primal[batch]
                      << ", status=" << status_vals[batch] << "\n";
        }
    }

    // Clean up
    {
        cudaFree(d_P_values);
        cudaFree(d_A_values);
        cudaFree(d_q);
        cudaFree(d_b);
        delete solver_ptr;
    }

    return 0;
}

File: packages/moreau-cpu/examples/rust/large_qp_batched.rs
#![allow(non_snake_case)]

//! Large Batched QP Benchmark
//!
//! Benchmarks solving many QP problems in parallel using CompiledSolver.
//! Matches the C++ large_qp_batched example for comparison.
//!
//! Usage: cargo run --example large_qp_batched --release [batch_size] [n] [m] [density]
//!
//! Default: batch_size=128, n=5000, m=2500, density=0.05

use moreau::solver::implementations::default::{BatchProblem, CompiledSolver, DefaultSettings};
use moreau::solver::SupportedConeT;
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use std::time::Instant;

fn main() {
    let args: Vec<String> = std::env::args().collect();

    // Default problem size (matches C++ defaults)
    let batch_size: usize = args.get(1).and_then(|s| s.parse().ok()).unwrap_or(128);
    let n: usize = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(5000);
    let m: usize = args.get(3).and_then(|s| s.parse().ok()).unwrap_or(2500);
    let density: f64 = args.get(4).and_then(|s| s.parse().ok()).unwrap_or(0.05);

    println!("Creating large batched QP problem:");
    println!("  Variables: {}", n);
    println!("  Constraints: {}", m);
    println!("  Batch size: {}", batch_size);
    println!("  Target density: {:.1}%", density * 100.0);

    let mut rng = StdRng::seed_from_u64(42);

    // P matrix: diagonal (n x n)
    let P_row_offsets: Vec<usize> = (0..=n).collect();
    let P_col_indices: Vec<usize> = (0..n).collect();
    let nnz_P = n;

    // A matrix: sparse random (m x n), with specified density
    let entries_per_row = std::cmp::max(1, std::cmp::min(n, (density * n as f64 + 0.5) as usize));
    let mut A_row_offsets: Vec<usize> = vec![0];
    let mut A_col_indices: Vec<usize> = Vec::new();

    for _ in 0..m {
        // Sample random columns
        let mut cols: Vec<usize> = (0..n).collect();
        // Fisher-Yates shuffle first entries_per_row elements
        for i in 0..entries_per_row {
            let j = rng.gen_range(i..n);
            cols.swap(i, j);
        }
        let mut selected: Vec<usize> = cols[..entries_per_row].to_vec();
        selected.sort();

        for col in selected {
            A_col_indices.push(col);
        }
        A_row_offsets.push(A_col_indices.len());
    }

    let nnz_A = A_col_indices.len();
    println!(
        "A matrix sparsity: {} nonzeros ({:.1}% dense)",
        nnz_A,
        100.0 * nnz_A as f64 / (n * m) as f64
    );

    // Cones: all nonnegative (simple box constraints)
    let cones: Vec<SupportedConeT<f64>> = vec![SupportedConeT::NonnegativeConeT(m)];

    // Settings
    let mut settings = DefaultSettings::default();
    settings.verbose = false;
    settings.max_iter = 50;

    println!("Creating solver...");

    let compiled_solver = CompiledSolver::new(
        n,
        m,
        &P_row_offsets,
        &P_col_indices,
        &A_row_offsets,
        &A_col_indices,
        &cones,
        settings,
        batch_size,
        false,
    )
    .expect("Failed to create batch solver");

    // Generate batch of problems with different values
    let mut problems = Vec::with_capacity(batch_size);

    for _ in 0..batch_size {
        // P values: random diagonal in [0.1, 2.0]
        let P_values: Vec<f64> = (0..nnz_P).map(|_| rng.gen_range(0.1..2.0)).collect();

        // q values: random in [-1, 1]
        let q: Vec<f64> = (0..n).map(|_| rng.gen_range(-1.0..1.0)).collect();

        // A values: random in [0.1, 2.0]
        let A_values: Vec<f64> = (0..nnz_A).map(|_| rng.gen_range(0.1..2.0)).collect();

        // b values: random in [-2, 2]
        let b: Vec<f64> = (0..m).map(|_| rng.gen_range(-2.0..2.0)).collect();

        problems.push(BatchProblem {
            P_values,
            q,
            A_values,
            b,
        });
    }

    println!("\nSolving {} problems in parallel...", batch_size);

    let start = Instant::now();
    let solutions = compiled_solver
        .solve_batch_parallel(&problems)
        .expect("Batch solve failed");
    let elapsed = start.elapsed();

    let total_ms = elapsed.as_secs_f64() * 1000.0;
    let per_problem_ms = total_ms / batch_size as f64;

    println!("\n=== BENCHMARK ===");
    println!("Total solve() time: {:.1} ms", total_ms);
    println!("Time per problem: {:.3} ms", per_problem_ms);
    println!("Throughput: {:.4} problems/second", 1000.0 / per_problem_ms);

    // Print first solution snippet
    println!("\n=== SOLUTION (first problem, first 10 values) ===");
    print!("x = [");
    for i in 0..std::cmp::min(10, n) {
        print!("{:.6}", solutions[0].x[i]);
        if i < 9 {
            print!(", ");
        }
    }
    println!(", ...]");

    // Count statuses
    let mut num_solved = 0;
    let mut num_almost = 0;
    let mut num_failed = 0;

    for sol in &solutions {
        match sol.status {
            moreau::solver::SolverStatus::Solved => num_solved += 1,
            moreau::solver::SolverStatus::AlmostSolved => num_almost += 1,
            _ => num_failed += 1,
        }
    }

    println!("\n=== BATCH RESULTS ===");
    println!("Solved: {} / {}", num_solved, batch_size);
    println!("Almost solved: {} / {}", num_almost, batch_size);
    println!("Failed: {} / {}", num_failed, batch_size);
}
